packages/expo/src/web/Image/AssetSourceResolver.ts

```diff
@@ -0,0 +1,98 @@
+import { Platform } from 'expo-core';
+import { UnavailabilityError } from 'expo-errors';
+import invariant from 'invariant';
+import { Dimensions } from 'react-native';
+
+import { getBasePath, PackagerAsset } from './assetPathUtils';
+
+export type ResolvedAssetSource = {
+  __packager_asset: boolean;
+  width?: number;
+  height?: number;
+  uri: string;
+  scale: number;
+};
+
+function getScale(): number {
+  return Dimensions.get('window').scale;
+}
+
+function getScaledAssetPath(asset): string {
+  const scale = AssetSourceResolver.pickScale(asset.scales, getScale());
+  const scaleSuffix = scale === 1 ? '' : '@' + scale + 'x';
+  const assetDir = getBasePath(asset);
+  return assetDir + '/' + asset.name + scaleSuffix + '.' + asset.type;
+}
+
+export default class AssetSourceResolver {
+  serverUrl?: string | null;
+  // where the jsbundle is being run from
+  jsbundleUrl?: string | null;
+  // the asset to resolve
+  asset: PackagerAsset;
+
+  constructor(
+    serverUrl: string | undefined | null,
+    jsbundleUrl: string | undefined | null,
+    asset: PackagerAsset
+  ) {
+    this.serverUrl = serverUrl;
+    this.jsbundleUrl = jsbundleUrl;
+    this.asset = asset;
+  }
+  isLoadedFromServer(): boolean {
+    return !!this.serverUrl;
+  }
+  isLoadedFromFileSystem(): boolean {
+    return !!(this.jsbundleUrl && this.jsbundleUrl.startsWith('file://'));
+  }
+  defaultAsset(): ResolvedAssetSource {
+    if (this.isLoadedFromServer()) {
+      return this.assetServerURL();
+    }
+
+    return this.scaledAssetURLNearBundle();
+  }
+  assetServerURL(): ResolvedAssetSource {
```

Comment:
Stuff like this seems unused -> maybe add a console.warn to each method and get an idea of what is called + grep to make a final decision about what needs to stay around.